src/bucket/BucketManagerImpl.cpp

```diff
@@ -599,21 +640,41 @@ BucketManagerImpl::getBucketIfExists(uint256 const& hash)
     return nullptr;
 }
 
-std::shared_ptr<Bucket>
+std::shared_ptr<LiveBucket>
+BucketManagerImpl::getLiveBucketByHash(uint256 const& hash)
+{
+    return getBucketByHash<LiveBucket>(hash);
+}
+
+std::shared_ptr<HotArchiveBucket>
+BucketManagerImpl::getHotArchiveBucketByHash(uint256 const& hash)
+{
+    return getBucketByHash<HotArchiveBucket>(hash);
+}
+
+template <class BucketT>
+std::shared_ptr<BucketT>
 BucketManagerImpl::getBucketByHash(uint256 const& hash)
 {
     ZoneScoped;
     std::lock_guard<std::recursive_mutex> lock(mBucketMutex);
     if (isZero(hash))
     {
-        return std::make_shared<Bucket>();
+        return std::make_shared<BucketT>();
     }
     auto i = mSharedBuckets.find(hash);
     if (i != mSharedBuckets.end())
     {
         CLOG_TRACE(Bucket, "BucketManager::getBucketByHash({}) found bucket {}",
                    binToHex(hash), i->second->getFilename());
-        return i->second;
+
+        // Because BucketManger has an impl class, no public templated functions
+        // can be declared. This means we have to manually enforce types via
+        // `getLiveBucketByHash` and `getHotBucketByHash`, leading to this ugly
+        // cast.
+        auto ret = std::dynamic_pointer_cast<BucketT>(i->second);
```

Comment:
Yeah unfortunately templated virtual functions are against the c++ spec, so the impl-class pattern is inconvenient here. I templated the inner BucketManagerImpl version, but we can only either have explicit return types with a messy dynamic cast, or just return a `BucketT` type in the `BucketManager` interface and have to cast on the caller side. I _think_ I might be able to get by with a `BucketT` level interface for futures and get rid of some of the constexpr branching, but we definitely need to be able to return `LiveBucket` explicitly in the   `getBucketByHash` case.

I'm fairly confident this is safe, since we have the Bucket type as part of meta so we're guaranteed to not have hash collision (except for empty buckets, but that should be fine). All that being said, I think separate `mSharedBuckets` would be useful so we could at least get rid of the dynamic cast, but we're stuck with the split `getLiveBucketByHash` interface I'm afraid.

Reply:
Hmm so I can't actually use a `BucketT` interface for futures either, but I can still have separate storage types in `BucketManager` to remove the cast. I think I can get rid of all the constexpr branching by changing the getBucket and getFutures interface to static getter that takes a `BucketManager&`. It's not a very elegant interface, but avoids removing the impl construct and removes most constexpr branching, so it's probably the least evil approach.

Reply:
note that this approach might be a bit annoying in tests, if we implement BucketList stub subclasses and such, but probably still worth exploring